### PR TITLE
LPS-134898

### PR DESCRIPTION
**Pull request by marianoalvarosaiz**

### CODE REVIEW

**Comment by tinatian** [comment]
@dantewang Please take this over.

Thanks.
Tina.

**Comment by dantewang** [comment]
@tinatian Got it.

File: modules/apps/expando/expando-test/src/testIntegration/java/com/liferay/expando/service/test/ExpandoValueLocalServiceTest.java [hunk 1]
```diff
@@ -309,15 +354,45 @@ public void testGetSerializableData() throws Exception {
...
+	}
+
 	private static final Log _log = LogFactoryUtil.getLog(
 		ExpandoValueLocalServiceTest.class);
 
 	private long _classNameId;
 	private Locale _enLocale;
 
+	@Inject
+	private ExpandoRowPersistence _expandoRowPersistence;
```
Comment (dantewang) [inline modules/apps/expando/expando-test/src/testIntegration/java/com/liferay/expando/service/test/ExpandoValueLocalServiceTest.java:388]: @marianoalvarosaiz These fields are injected but not used in your test code. Are they needed?

**Comment by dantewang** [comment]
@marianoalvarosaiz I ran your test without your fix, and got the following exception:

```
com.liferay.expando.service.test.ExpandoValueLocalServiceTest > testAddValuesWithFlushInBetween FAILED
    com.liferay.portal.kernel.exception.SystemException: com.liferay.portal.kernel.dao.orm.ORMException: javax.persistence.OptimisticLockException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [com.liferay.portlet.expando.model.impl.ExpandoRowImpl#43541]
    	at com.liferay.portal.kernel.service.persistence.impl.BasePersistenceImpl.processException(BasePersistenceImpl.java:639)
    	at com.liferay.portlet.expando.service.persistence.impl.ExpandoRowPersistenceImpl.updateImpl(ExpandoRowPersistenceImpl.java:1616)
    	at com.liferay.portlet.expando.service.persistence.impl.ExpandoRowPersistenceImpl.updateImpl(ExpandoRowPersistenceImpl.java:75)
    	at com.liferay.portal.kernel.service.persistence.impl.BasePersistenceImpl.update(BasePersistenceImpl.java:786)
    	at com.liferay.portlet.expando.service.impl.ExpandoValueLocalServiceImpl.doAddValue(ExpandoValueLocalServiceImpl.java:1745)
    	at com.liferay.portlet.expando.service.impl.ExpandoValueLocalServiceImpl.addValue(ExpandoValueLocalServiceImpl.java:77)
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.lang.reflect.Method.invoke(Method.java:498)
    	at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:50)
    	at com.liferay.portal.spring.transaction.TransactionInterceptor.invoke(TransactionInterceptor.java:69)
    	at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:57)
    	at com.liferay.portal.spring.aop.AopInvocationHandler.invoke(AopInvocationHandler.java:49)
    	at com.sun.proxy.$Proxy177.addValue(Unknown Source)
    	at com.liferay.portlet.expando.service.impl.ExpandoValueLocalServiceImpl.addValue(ExpandoValueLocalServiceImpl.java:651)
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.lang.reflect.Method.invoke(Method.java:498)
    	at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:50)
    	at com.liferay.portal.spring.transaction.TransactionInterceptor.invoke(TransactionInterceptor.java:69)
    	at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:57)
    	at com.liferay.portal.spring.aop.AopInvocationHandler.invoke(AopInvocationHandler.java:49)
    	at com.sun.proxy.$Proxy177.addValue(Unknown Source)
    	at com.liferay.portlet.expando.service.impl.ExpandoValueLocalServiceImpl.addValue(ExpandoValueLocalServiceImpl.java:577)
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.lang.reflect.Method.invoke(Method.java:498)
    	at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:50)
    	at com.liferay.portal.spring.transaction.TransactionInterceptor.invoke(TransactionInterceptor.java:69)
    	at com.liferay.portal.spring.aop.AopMethodInvocationImpl.proceed(AopMethodInvocationImpl.java:57)
    	at com.liferay.portal.spring.aop.AopInvocationHandler.invoke(AopInvocationHandler.java:49)
    	at com.sun.proxy.$Proxy177.addValue(Unknown Source)
    	at com.liferay.expando.kernel.service.ExpandoValueLocalServiceUtil.addValue(ExpandoValueLocalServiceUtil.java:225)
    	at com.liferay.portlet.expando.util.test.ExpandoTestUtil.addValue(ExpandoTestUtil.java:65)
    	at com.liferay.expando.service.test.ExpandoValueLocalServiceTest.lambda$testAddValuesWithFlushInBetween$0(ExpandoValueLocalServiceTest.java:207)
    	at com.liferay.portal.spring.transaction.BaseTransactionExecutor.execute(BaseTransactionExecutor.java:37)
    	at com.liferay.portal.spring.transaction.TransactionInvokerImpl.invoke(TransactionInvokerImpl.java:39)
    	at com.liferay.portal.kernel.transaction.TransactionInvokerUtil.invoke(TransactionInvokerUtil.java:28)
    	at com.liferay.expando.service.test.ExpandoValueLocalServiceTest.testAddValuesWithFlushInBetween(ExpandoValueLocalServiceTest.java:190)
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.lang.reflect.Method.invoke(Method.java:498)
    	at com.liferay.arquillian.extension.junit.bridge.server.TestExecutorRunnable$3.evaluate(TestExecutorRunnable.java:353)
    	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
    	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
    	at com.liferay.arquillian.extension.junit.bridge.server.TestExecutorRunnable$1.evaluate(TestExecutorRunnable.java:237)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
    	at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:59)
    	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
    	at com.liferay.arquillian.extension.junit.bridge.server.TestExecutorRunnable._execute(TestExecutorRunnable.java:316)
    	at com.liferay.arquillian.extension.junit.bridge.server.TestExecutorRunnable.run(TestExecutorRunnable.java:95)
    	at java.lang.Thread.run(Thread.java:748)
    Caused by: com.liferay.portal.kernel.dao.orm.ORMException: javax.persistence.OptimisticLockException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [com.liferay.portlet.expando.model.impl.ExpandoRowImpl#43541]
    	at com.liferay.portal.dao.orm.hibernate.ExceptionTranslator.translate(ExceptionTranslator.java:78)
    	at com.liferay.portal.dao.orm.hibernate.SessionImpl.merge(SessionImpl.java:346)
    	at com.liferay.portal.kernel.dao.orm.SessionWrapper.merge(SessionWrapper.java:160)
    	at com.liferay.portlet.expando.service.persistence.impl.ExpandoRowPersistenceImpl.updateImpl(ExpandoRowPersistenceImpl.java:1612)
    	... 68 more
    Caused by: javax.persistence.OptimisticLockException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [com.liferay.portlet.expando.model.impl.ExpandoRowImpl#43541]
    	at org.hibernate.internal.ExceptionConverterImpl.wrapStaleStateException(ExceptionConverterImpl.java:223)
    	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:93)
    	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
    	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:188)
    	at org.hibernate.internal.SessionImpl.fireMerge(SessionImpl.java:840)
    	at org.hibernate.internal.SessionImpl.merge(SessionImpl.java:816)
    	at com.liferay.portal.dao.orm.hibernate.SessionImpl.merge(SessionImpl.java:343)
    	... 70 more
    Caused by: org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [com.liferay.portlet.expando.model.impl.ExpandoRowImpl#43541]
    	at org.hibernate.event.internal.DefaultMergeEventListener.entityIsDetached(DefaultMergeEventListener.java:341)
    	at org.hibernate.event.internal.DefaultMergeEventListener.onMerge(DefaultMergeEventListener.java:172)
    	at org.hibernate.event.internal.DefaultMergeEventListener.onMerge(DefaultMergeEventListener.java:70)
    	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:107)
    	at org.hibernate.internal.SessionImpl.fireMerge(SessionImpl.java:829)
    	... 72 more

```

Is this the one I should expect?

However, with this test, the root cause is still unclear. How does the model listener affect the update? Could you see if it's possible to make the test more self-explaining, or add some detailed explanation to the ticket?

File: modules/apps/expando/expando-test/src/testIntegration/java/com/liferay/expando/service/test/ExpandoValueLocalServiceTest.java (hunk 1 above)
Comment (marianoalvarosaiz) [reply modules/apps/expando/expando-test/src/testIntegration/java/com/liferay/expando/service/test/ExpandoValueLocalServiceTest.java:388]: @dantewang You're right, I forgot to remove it from the previous test.

**Comment by marianoalvarosaiz** [comment]
> Is this the one I should expect?
> However, with this test, the root cause is still unclear. How does the model listener affect the update? Could you see if it's possible to make the test more self-explaining, or add some detailed explanation to the ticket?

@dantewang  Yes, that's correct.

The problem we are trying to fix, or minimize its probability to occur, is exactly that, that the row retrieved can be stale. From the moment we retrieve the _row_ the first time in the method until the moment we perform the update over the _row_ other operations (or flushing) can occur since we are, in the middle, executing other instructions, including an update over the _value_. If we retrieve the row immediately before updating it the probability of the row being stale gets reduced and it doesn't affect performance since it should be already cached.

In the end it's just a matter of making the time between the get and the update over the _row_ as short as possible and not having an update over a different model in the middle.

File: modules/apps/expando/expando-test/src/testIntegration/java/com/liferay/expando/service/test/ExpandoValueLocalServiceTest.java (hunk 1 above)
Comment (marianoalvarosaiz) [reply modules/apps/expando/expando-test/src/testIntegration/java/com/liferay/expando/service/test/ExpandoValueLocalServiceTest.java:388]: I can resend a new pull without the field once the review is completed.

**Comment by dantewang** [comment]
@marianoalvarosaiz When you resend the PR, could you highlight (maybe by comment or other ways) in your test where the update by one affects the retrieve and the update by the other one (how your test forms such a situation)? This is the major problem to me -- the test works fine but it's a bit hard to understand when reading the code.

File: modules/apps/expando/expando-test/src/testIntegration/java/com/liferay/expando/service/test/ExpandoValueLocalServiceTest.java (hunk 1 above)
Comment (dantewang) [reply modules/apps/expando/expando-test/src/testIntegration/java/com/liferay/expando/service/test/ExpandoValueLocalServiceTest.java:388]: @marianoalvarosaiz Thanks

**Comment by marianoalvarosaiz** [comment]
@dantewang something like this would be fine? https://github.com/marianoalvarosaiz/liferay-portal/pull/835

**Comment by tinatian** [comment]
ci:close

**Comment by dantewang** [comment]
@marianoalvarosaiz Yes, please resend

**Comment by marianoalvarosaiz** [comment]
@dantewang see https://github.com/liferay-core-infra/liferay-portal/pull/945

**Pull request closed without merging**